File: clipboard_adv.rb
# clipboard_adv.rb
require 'json'
require 'time'
require 'thread'

begin
  require 'clipboard'
rescue LoadError
  puts "Please install: gem install clipboard"
  exit(1)
end

HISTORY_FILE = 'clipboard_adv_history.json'
MAX_HISTORY = 200
DEFAULT_CATEGORY = 'other'
$history = []
$last_clipboard = ''
$running = true

def load_history
  $history = JSON.parse(File.read(HISTORY_FILE)) rescue []
end

def save_history
  File.write(HISTORY_FILE, JSON.pretty_generate($history))
end

def add_entry(text)
  return if text.nil? || text.strip.empty?
  return if $history.first && $history.first['text'] == text
  entry = {
    'id' => Time.now.to_i,
    'text' => text,
    'timestamp' => Time.now.iso8601,
    'category' => DEFAULT_CATEGORY,
    'pinned' => false,
    'source_app' => 'unknown',
    'usage_count' => 0
  }
  $history.unshift(entry)
  if $history.size > MAX_HISTORY
    # remove oldest non-pinned
    $history.each_with_index do |e, i|
      next if i == 0
      unless e['pinned']
        $history.delete_at(i)
        break
      end
    end
    $history.pop if $history.size > MAX_HISTORY
  end
  save_history
end

def truncate(s, n)
  s.length > n ? s[0...n] + '...' : s
end

def show_history
  if $history.empty?
    puts 'History is empty.'
    return
  end
  $history.each_with_index do |e, i|
    pin = e['pinned'] ? '★ ' : ''
    puts "[#{i+1}] #{pin}#{e['category']}: #{truncate(e['text'], 50)}  (#{e['timestamp']})  [used: #{e['usage_count']}]"
  end
end

def search_history(query)
  is_regex = false
  case_sensitive = true
  pattern = query
  if query.start_with?('/') && (query.end_with?('/') || query.end_with?('/i'))
    is_regex = true
    if query.end_with?('/i')
      pattern = query[1...-2]
      case_sensitive = false
    else
      pattern = query[1...-1]
      case_sensitive = true
    end
  end
  results = []
  $history.each do |e|
    if is_regex
      begin
        flags = case_sensitive ? 0 : Regexp::IGNORECASE
        if Regexp.new(pattern, flags) =~ e['text']
          results << e
        end
      rescue => ex
        puts "Invalid regex."
        return
      end
    else
      if e['text'].downcase.include?(query.downcase)
        results << e
      end
    end
  end
  if results.empty?
    puts 'No matches found.'
    return
  end
  results.each_with_index do |e, i|
    pin = e['pinned'] ? '★ ' : ''
    puts "[#{i+1}] #{pin}#{e['category']}: #{truncate(e['text'], 50)}  (#{e['timestamp']})  [used: #{e['usage_count']}]"
  end
end

def paste_entry(index)
  if index < 1 || index > $history.size
    puts 'Invalid index.'
    return
  end
  e = $history[index-1]
  Clipboard.copy(e['text'])
  e['usage_count'] += 1
  save_history
  puts "Copied to clipboard: #{truncate(e['text'], 50)}"
end

def delete_entry(index)
  if index < 1 || index > $history.size
    puts 'Invalid index.'
    return
  end
  removed = $history.delete_at(index-1)
  save_history
  puts "Deleted: #{truncate(removed['text'], 50)}"
end

def toggle_pin(index)
  if index < 1 || index > $history.size
    puts 'Invalid index.'
    return
  end
  e = $history[index-1]
  e['pinned'] = !e['pinned']
  save_history
  status = e['pinned'] ? 'pinned' : 'unpinned'
  puts "Entry #{status}: #{truncate(e['text'], 50)}"
end

def set_category(index, category)
  if index < 1 || index > $history.size
    puts 'Invalid index.'
    return
  end
  e = $history[index-1]
  e['category'] = category
  save_history
  puts "Category updated: #{truncate(e['text'], 50)} -> #{category}"
end

def show_stats
  total = $history.size
  if total == 0
    puts 'No entries.'
    return
  end
  categories = {}
  pinned_count = 0
  total_usage = 0
  most_used = 0
  ages = []
  now = Time.now
  $history.each do |e|
    categories[e['category']] = categories.fetch(e['category'], 0) + 1
    pinned_count += 1 if e['pinned']
    total_usage += e['usage_count']
    most_used = e['usage_count'] if e['usage_count'] > most_used
    begin
      ts = Time.parse(e['timestamp'])
      ages << (now - ts) / (60 * 60 * 24)
    rescue
    end
  end
  avg_age = ages.empty? ? 0 : ages.sum / ages.size
  puts 'Statistics:'
  puts "  Total entries: #{total}"
  puts "  Pinned: #{pinned_count}"
  puts "  Categories: #{categories.map { |k,v| "#{k}:#{v}" }.join(', ')}"
  puts "  Most used count: #{most_used}"
  puts "  Total usage: #{total_usage}"
  puts "  Average age: #{'%.1f' % avg_age} days"
end

def cleanup(days)
  if days <= 0
    puts 'Days must be positive.'
    return
  end
  now = Time.now
  removed = 0
  kept = []
  $history.each do |e|
    if e['pinned']
      kept << e
      next
    end
    begin
      ts = Time.parse(e['timestamp'])
      if (now - ts) / (60 * 60 * 24) > days
        removed += 1
      else
        kept << e
      end
    rescue
      kept << e
    end
  end
  $history = kept
  save_history
  puts "Removed #{removed} entries older than #{days} days."
end

def export_json(filename)
  File.write(filename, JSON.pretty_generate($history))
  puts "Exported to #{filename}"
end

def export_csv(filename)
  CSV.open(filename, 'w') do |csv|
    csv << ['Text', 'Category', 'Timestamp', 'Pinned', 'UsageCount']
    $history.each do |e|
      csv << [e['text'], e['category'], e['timestamp'], e['pinned'], e['usage_count']]
    end
  end
  puts "Exported to #{filename}"
end

def import_json(filename)
  data = JSON.parse(File.read(filename))
  if data.is_a?(Array)
    $history = data
    save_history
    puts "Imported #{$history.size} entries from #{filename}"
  else
    puts 'Invalid format.'
  end
rescue => e
  puts "Error: #{e.message}"
end

def import_csv(filename)
  require 'csv'
  imported = []
  CSV.foreach(filename, headers: true) do |row|
    entry = {
      'text' => row['Text'],
      'category' => row['Category'] || DEFAULT_CATEGORY,
      'timestamp' => row['Timestamp'] || Time.now.iso8601,
      'pinned' => row['Pinned'] == 'true',
      'usage_count' => row['UsageCount'].to_i,
      'id' => Time.now.to_i + imported.size,
      'source_app' => 'imported'
    }
    imported << entry
  end
  $history = imported
  save_history
  puts "Imported #{$history.size} entries from #{filename}"
rescue => e
  puts "Error: #{e.message}"
end

def main
  load_history
  $last_clipboard = Clipboard.paste

  Thread.new do
    while $running
      begin
        current = Clipboard.paste
        if current && current != $last_clipboard && !current.empty?
          $last_clipboard = current
          add_entry(current)
        end
        sleep 0.5
      rescue
      end
    end
  end

  puts "=== Advanced Clipboard Manager ==="
  puts "Monitoring clipboard..."
  puts "Commands: 1=history, 2=add, 3=search, 4=paste, 5=delete, 6=pin, 7=category, 8=stats, 9=cleanup, 10=export, 11=import, 12=exit"

  loop do
    print "\n> "
    cmd = gets.chomp.strip
    case cmd
    when '1'
      show_history
    when '2'
      text = Clipboard.paste
      add_entry(text)
      puts "Added: #{truncate(text, 50)}"
    when '3'
      print 'Search query (regex supported: /pattern/i): '
      query = gets.chomp.strip
      search_history(query) unless query.empty?
    when '4'
      print 'Enter index: '
      idx = gets.to_i
      paste_entry(idx)
    when '5'
      print 'Enter index: '
      idx = gets.to_i
      delete_entry(idx)
    when '6'
      print 'Enter index: '
      idx = gets.to_i
      toggle_pin(idx)
    when '7'
      print 'Enter index: '
      idx = gets.to_i
      print 'Enter category (code/link/note/quote/other): '
      cat = gets.chomp.strip
      cat = DEFAULT_CATEGORY if cat.empty?
      set_category(idx, cat)
    when '8'
      show_stats
    when '9'
      print 'Remove entries older than N days: '
      days = gets.to_i
      cleanup(days)
    when '10'
      print 'Export format (json/csv): '
      fmt = gets.chomp.strip.downcase
      print "Filename (default: export.#{fmt}): "
      fname = gets.chomp.strip
      fname = "export.#{fmt}" if fname.empty?
      if fmt == 'json'
        export_json(fname)
      elsif fmt == 'csv'
        export_csv(fname)
      else
        puts 'Unknown format.'
      end
    when '11'
      print 'Import format (json/csv): '
      fmt = gets.chomp.strip.downcase
      print 'Filename: '
      fname = gets.chomp.strip
      if fname.empty?
        puts 'Filename required.'
        next
      end
      if fmt == 'json'
        import_json(fname)
      elsif fmt == 'csv'
        import_csv(fname)
      else
        puts 'Unknown format.'
      end
    when '12'
      puts 'Goodbye!'
      $running = false
      break
    else
      puts 'Invalid command.'
    end
  end
end

main if __FILE__ == $0

File: clipboard_adv.py
# clipboard_adv.py
import json
import time
import os
import re
from datetime import datetime, timedelta
import threading
import sys

try:
    import pyperclip
    import keyboard
except ImportError:
    print("Please install: pip install pyperclip keyboard")
    sys.exit(1)

HISTORY_FILE = "clipboard_adv_history.json"
MAX_HISTORY = 200
DEFAULT_CATEGORY = "other"
history = []
last_clipboard = ""
running = True

def load_history():
    global history
    try:
        with open(HISTORY_FILE, 'r', encoding='utf-8') as f:
            history = json.load(f)
    except (FileNotFoundError, json.JSONDecodeError):
        history = []

def save_history():
    with open(HISTORY_FILE, 'w', encoding='utf-8') as f:
        json.dump(history, f, indent=2, ensure_ascii=False)

def add_entry(text, source_app=""):
    if not text or text.strip() == "":
        return
    # Avoid duplicates (last entry)
    if history and history[0]["text"] == text:
        return
    entry = {
        "id": int(time.time()),
        "text": text,
        "timestamp": datetime.now().isoformat(),
        "category": DEFAULT_CATEGORY,
        "pinned": False,
        "source_app": source_app,
        "usage_count": 0
    }
    history.insert(0, entry)
    if len(history) > MAX_HISTORY:
        # Remove oldest non-pinned entries
        non_pinned = [i for i, e in enumerate(history) if not e["pinned"]]
        if non_pinned:
            del history[non_pinned[-1]]
        else:
            history.pop()
    save_history()

def monitor_clipboard():
    global last_clipboard
    while running:
        try:
            current = pyperclip.paste()
            if current and current != last_clipboard:
                last_clipboard = current
                # Try to get source app (not easily cross-platform; placeholder)
                add_entry(current, "unknown")
        except:
            pass
        time.sleep(0.5)

def show_history():
    if not history:
        print("History is empty.")
        return
    for i, entry in enumerate(history, 1):
        pin = "★ " if entry.get("pinned", False) else ""
        cat = entry.get("category", DEFAULT_CATEGORY)
        used = entry.get("usage_count", 0)
        print(f"[{i}] {pin}{cat}: {entry['text'][:50]}  ({entry['timestamp']})  [used: {used}]")

def search_history(query):
    # Check if query is a regex pattern
    is_regex = False
    case_sensitive = True
    pattern = query
    if query.startswith('/') and (query.endswith('/') or query.endswith('/i')):
        is_regex = True
        if query.endswith('/i'):
            pattern = query[1:-2]
            case_sensitive = False
        else:
            pattern = query[1:-1]
            case_sensitive = True
    results = []
    for entry in history:
        text = entry["text"]
        if is_regex:
            try:
                flags = 0 if case_sensitive else re.IGNORECASE
                if re.search(pattern, text, flags):
                    results.append(entry)
            except re.error:
                print("Invalid regex.")
                return
        else:
            if query.lower() in text.lower():
                results.append(entry)
    if not results:
        print("No matches found.")
        return
    for i, entry in enumerate(results, 1):
        pin = "★ " if entry.get("pinned", False) else ""
        cat = entry.get("category", DEFAULT_CATEGORY)
        used = entry.get("usage_count", 0)
        print(f"[{i}] {pin}{cat}: {entry['text'][:50]}  ({entry['timestamp']})  [used: {used}]")

def paste_entry(index):
    if not history:
        print("History is empty.")
        return
    if 1 <= index <= len(history):
        entry = history[index-1]
        pyperclip.copy(entry["text"])
        # Increment usage count
        entry["usage_count"] = entry.get("usage_count", 0) + 1
        save_history()
        print(f"Copied to clipboard: {entry['text'][:50]}")
    else:
        print("Invalid index.")

def delete_entry(index):
    if 1 <= index <= len(history):
        removed = history.pop(index-1)
        save_history()
        print(f"Deleted: {removed['text'][:50]}")
    else:
        print("Invalid index.")

def toggle_pin(index):
    if 1 <= index <= len(history):
        entry = history[index-1]
        entry["pinned"] = not entry.get("pinned", False)
        save_history()
        status = "pinned" if entry["pinned"] else "unpinned"
        print(f"Entry {status}: {entry['text'][:50]}")
    else:
        print("Invalid index.")

def set_category(index, category):
    if 1 <= index <= len(history):
        entry = history[index-1]
        entry["category"] = category
        save_history()
        print(f"Category updated: {entry['text'][:50]} -> {category}")
    else:
        print("Invalid index.")

def show_stats():
    total = len(history)
    if total == 0:
        print("No entries.")
        return
    categories = {}
    pinned_count = 0
    usage_counts = [e.get("usage_count", 0) for e in history]
    most_used = max(usage_counts) if usage_counts else 0
    total_usage = sum(usage_counts)
    for e in history:
        cat = e.get("category", DEFAULT_CATEGORY)
        categories[cat] = categories.get(cat, 0) + 1
        if e.get("pinned", False):
            pinned_count += 1
    # Average age
    now = datetime.now()
    ages = []
    for e in history:
        try:
            ts = datetime.fromisoformat(e["timestamp"])
            ages.append((now - ts).days)
        except:
            pass
    avg_age = sum(ages)/len(ages) if ages else 0
    print(f"Statistics:")
    print(f"  Total entries: {total}")
    print(f"  Pinned: {pinned_count}")
    print(f"  Categories: {', '.join(f'{k}:{v}' for k,v in categories.items())}")
    print(f"  Most used count: {most_used}")
    print(f"  Total usage: {total_usage}")
    print(f"  Average age: {avg_age:.1f} days")

def cleanup(days):
    if days <= 0:
        print("Days must be positive.")
        return
    now = datetime.now()
    removed = 0
    to_keep = []
    for e in history:
        if e.get("pinned", False):
            to_keep.append(e)
            continue
        try:
            ts = datetime.fromisoformat(e["timestamp"])
            if (now - ts).days > days:
                removed += 1
            else:
                to_keep.append(e)
        except:
            to_keep.append(e)
    global history
    history = to_keep
    save_history()
    print(f"Removed {removed} entries older than {days} days.")

def export_json(filename):
    with open(filename, 'w', encoding='utf-8') as f:
        json.dump(history, f, indent=2, ensure_ascii=False)
    print(f"Exported to {filename}")

def export_csv(filename):
    import csv
    with open(filename, 'w', newline='', encoding='utf-8') as f:
        writer = csv.writer(f)
        writer.writerow(["Text", "Category", "Timestamp", "Pinned", "UsageCount"])
        for e in history:
            writer.writerow([e["text"], e.get("category", DEFAULT_CATEGORY),
                             e["timestamp"], e.get("pinned", False), e.get("usage_count", 0)])
    print(f"Exported to {filename}")

def import_json(filename):
    global history
    try:
        with open(filename, 'r', encoding='utf-8') as f:
            imported = json.load(f)
        if isinstance(imported, list):
            history = imported
            save_history()
            print(f"Imported {len(history)} entries from {filename}")
        else:
            print("Invalid format: expected list.")
    except Exception as e:
        print(f"Error: {e}")

def import_csv(filename):
    global history
    import csv
    try:
        with open(filename, 'r', encoding='utf-8') as f:
            reader = csv.DictReader(f)
            new_history = []
            for row in reader:
                entry = {
                    "text": row["Text"],
                    "category": row.get("Category", DEFAULT_CATEGORY),
                    "timestamp": row.get("Timestamp", datetime.now().isoformat()),
                    "pinned": row.get("Pinned", "False").lower() == "true",
                    "usage_count": int(row.get("UsageCount", 0)),
                    "id": int(time.time() + len(new_history))
                }
                new_history.append(entry)
            history = new_history
            save_history()
            print(f"Imported {len(history)} entries from {filename}")
    except Exception as e:
        print(f"Error: {e}")

def main():
    global running
    load_history()
    global last_clipboard
    last_clipboard = pyperclip.paste()

    monitor_thread = threading.Thread(target=monitor_clipboard, daemon=True)
    monitor_thread.start()

    print("=== Advanced Clipboard Manager ===")
    print("Monitoring clipboard... (Ctrl+C to stop)")
    print("Commands: 1=history, 2=add, 3=search, 4=paste, 5=delete, 6=pin, 7=category, 8=stats, 9=cleanup, 10=export, 11=import, 12=exit")

    try:
        while True:
            cmd = input("\n> ").strip()
            if cmd == "1":
                show_history()
            elif cmd == "2":
                text = pyperclip.paste()
                add_entry(text)
                print(f"Added: {text[:50]}")
            elif cmd == "3":
                query = input("Search query (regex supported: /pattern/i): ").strip()
                if query:
                    search_history(query)
            elif cmd == "4":
                idx = input("Enter index: ").strip()
                if idx.isdigit():
                    paste_entry(int(idx))
            elif cmd == "5":
                idx = input("Enter index: ").strip()
                if idx.isdigit():
                    delete_entry(int(idx))
            elif cmd == "6":
                idx = input("Enter index: ").strip()
                if idx.isdigit():
                    toggle_pin(int(idx))
            elif cmd == "7":
                idx = input("Enter index: ").strip()
                if not idx.isdigit():
                    print("Invalid index.")
                    continue
                cat = input("Enter category (code/link/note/quote/other): ").strip() or DEFAULT_CATEGORY
                set_category(int(idx), cat)
            elif cmd == "8":
                show_stats()
            elif cmd == "9":
                days = input("Remove entries older than N days: ").strip()
                if days.isdigit():
                    cleanup(int(days))
                else:
                    print("Invalid number.")
            elif cmd == "10":
                fmt = input("Export format (json/csv): ").strip().lower()
                fname = input("Filename (default: export.{}): ".format(fmt)).strip()
                if not fname:
                    fname = f"export.{fmt}"
                if fmt == "json":
                    export_json(fname)
                elif fmt == "csv":
                    export_csv(fname)
                else:
                    print("Unknown format.")
            elif cmd == "11":
                fmt = input("Import format (json/csv): ").strip().lower()
                fname = input("Filename: ").strip()
                if not fname:
                    print("Filename required.")
                    continue
                if fmt == "json":
                    import_json(fname)
                elif fmt == "csv":
                    import_csv(fname)
                else:
                    print("Unknown format.")
            elif cmd == "12":
                print("Goodbye!")
                running = False
                break
            else:
                print("Invalid command.")
    except KeyboardInterrupt:
        print("\nGoodbye!")
        running = False

if __name__ == "__main__":
    main()
